torch/csrc/jit/codegen/cuda/scheduler_registry.cpp

```diff
@@ -0,0 +1,320 @@
+    for (auto expr : fusion->exprs()) {
+      if (auto red = dynamic_cast<ReductionOp*>(expr)) {
+        if (!isTrivialReduction(red)) {
+          return red->out()->as<TensorView>();
+        }
+      }
+    }
+    TORCH_INTERNAL_ASSERT(false, "unreachable");
+    return nullptr;
+  }
+};
+
+class PointWiseScheduler : public SchedulerEntry {
+ public:
+  explicit PointWiseScheduler(Fusion* fusion)
+      : SchedulerEntry(ScheduleHeuristic::PointWise, false) {}
+
+  static bool canSchedule(Fusion* fusion) {
+    auto red_ops = findReductionOps(fusion);
+    return red_ops.empty();
+  }
+
+  void schedule(Fusion* fusion) override {
+    scheduleFusion(fusion);
+  }
+};
+
+namespace {
+
+// duplicated from Benchmark/utils.h
+static void analyzeFusion(
+    Fusion* fusion,
+    std::vector<TensorView*>& reduction_tv,
+    std::vector<TensorView*>& other_tv) {
+  auto all_values = DependencyCheck::getAllValsBetween(
+      {fusion->inputs().begin(), fusion->inputs().end()}, fusion->outputs());
+
+  for (auto tv : ir_utils::filterByType<TensorView>(all_values)) {
+    if (tv->hasReduction() && !fusion->hasInput(tv)) {
+      reduction_tv.push_back(tv);
+    } else if (!fusion->hasInput(tv)) {
+      other_tv.push_back(tv);
+    }
+  }
+}
+
+} // namespace
+class NormalizationScheduler : public SchedulerEntry {
+ public:
+  explicit NormalizationScheduler(Fusion* fusion, ExpressionEvaluator& ee)
+      : SchedulerEntry(ScheduleHeuristic::Normalization, true) {
+    getHeuristics(fusion, ee);
+  }
+
+  void schedule(Fusion* fusion) override {
+    std::vector<TensorView*> reduction_tensors;
+    std::vector<TensorView*> other_tensors;
+    analyzeFusion(fusion, reduction_tensors, other_tensors);
+    scheduleNormalization(fusion, rparams_, reduction_tensors, other_tensors);
+  }
+
+  static bool canSchedule(Fusion* fusion) {
+    auto red_ops = findReductionOps(fusion);
+
+    if (red_ops.size() < 2) {
+      // Use single reduction or pointwise logic
+      return false;
+    }
+    // Before examining the reduction axes want to quickly
+    //   check the reductions have the same axis width
+    //   to avoid building root domain map in easier cases
+    bool valid_axis_count = false;
+    size_t axis_count = 0;
+    auto reduction_root = [](ReductionOp* rop) {
+      return rop->out()->as<TensorView>()->getRootDomain();
+    };
+    for (auto red : red_ops) {
+      if (!valid_axis_count) {
+        valid_axis_count = true;
+        axis_count = reduction_root(red).size();
+      } else {
+        if (reduction_root(red).size() != axis_count) {
+          return false;
+        }
+      }
+    }
+
+    // Use root domain map to check the reduction ops have the same axes
+    FusionGuard fg(fusion);
+    ComputeAtRootDomainMap root_map;
+    root_map.build(true);
+
+    // red_ops.size()>1 checked before
+    for (size_t it = 1; it < red_ops.size(); it++) {
+      if (!checkEquivalence(red_ops[it - 1], red_ops[it], root_map)) {
```

Comment:
Nice use of root map CC @rdspring1 to review.

Reply:
LGTM